ports/stm/supervisor/internal_flash.c

```diff
@@ -113,6 +250,9 @@ static int32_t convert_block_to_flash_addr(uint32_t block) {
 }
 
 mp_uint_t supervisor_flash_read_blocks(uint8_t *dest, uint32_t block, uint32_t num_blocks) {
+    // Must write out anything in cache before trying to read.
```

Comment:
Why can't you read it from the cache?

Reply:
copied this line from nrf, figured it was policy. Not sure why it's there, will think about whether I'm missing something.

Reply:
Thinking about it, I feel like reading from the flash would have the potential to be problematic, because it means reading doesn't represent the actual contents of the flash. It'd also break the implementation from the Atmel and NRF ports which both have read relaying the actual physical state of flash. @dhalbert what do you think?

Reply:
The `atmel-samd` port doesn't have that flush, but it doesn't have a cache for internal flash, so the cache flush routines do nothing. A call to `supervisor_flash_flush()` should probably be added in the `atmel-samd` code to prevent a bug from cropping up if we did add a cache.

I added the flush to the `nrf` port in #1661 because the reads were ignoring the cache, which was a bug. In retrospect I guess I could have read from the cache instead. There's a background task that flushes the cache periodically. At the time I was really worried about filesystem integrity, because we were having severe filesystem problems on nRF boards with internal-flash-only filesystems.

Reply:
I guess the advantages in each direction would be
- read from cache) performance improvements, especially between rapidly interspersed read/write/read/write commands
- read from flash) avoids errors if a read is done to verify physical flash state, such as checking for shutdown readiness or something. 

Note that on the H7, the flash sectors are so huge that the filesystem will never carry over between them, so the background flush is the only time the cache will ever get written to disc.

Reply:
The huge flash sectors really point toward reading from the cache instead of flushing, since a write is expensive and wearing.